File: cpp/test/computed_fields_test.cc
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

#include <filesystem>
#include <type_traits>
#include <vector>

#include <gtest/gtest.h>

#include "generated/types.h"

using namespace test_model;

namespace {

TEST(ComputedFieldsTests, AccessFields) {
  RecordWithComputedFields r;

  r.int_field = 42;
  EXPECT_EQ(r.AccessIntField(), r.int_field);

  r.string_field = "hello";
  EXPECT_EQ(r.AccessStringField(), r.string_field);

  r.tuple_field = {2, 3};
  EXPECT_EQ(r.AccessTupleField(), r.tuple_field);
  EXPECT_EQ(r.AccessNestedTupleField(), r.tuple_field.v2);

  r.array_field = {{1, 2, 3}, {4, 5, 6}};
  EXPECT_EQ(r.AccessArrayField(), r.array_field);
  EXPECT_EQ(r.AccessArrayFieldElement(), yardl::at(r.array_field, 0, 1));
  EXPECT_EQ(r.AccessArrayFieldElementByName(), yardl::at(r.array_field, 0, 1));

  EXPECT_EQ(r.AccessOtherComputedField(), r.AccessIntField());

  r.vector_of_vectors_field = {{1, 2, 3}, {4, 5, 6}};
  EXPECT_EQ(r.AccessVectorOfVectorsField(), r.vector_of_vectors_field[1][2]);

  r.map_field = {{"hello", "world"}, {"world", "bye"}};
  EXPECT_EQ(r.AccessMap(), r.map_field);
  EXPECT_EQ(r.AccessMapEntry(), "world");
  EXPECT_EQ(r.AccessMapEntryWithComputedField(), "world");
  EXPECT_EQ(r.AccessMapEntryWithComputedFieldNested(), "bye");
  EXPECT_THROW(r.AccessMissingMapEntry(), std::out_of_range);
}

TEST(ComputedFieldsTest, Literals) {
  RecordWithComputedFields r;

  static_assert(std::is_same_v<decltype(r.IntLiteral()), uint8_t>);
  static_assert(std::is_same_v<decltype(r.LargeNegativeInt64Literal()), int64_t>);
  static_assert(std::is_same_v<decltype(r.LargeUInt64Literal()), uint64_t>);

  ASSERT_EQ(r.StringLiteral(), "hello");
  ASSERT_EQ(r.StringLiteral2(), "hello");
  ASSERT_EQ(r.StringLiteral3(), "hello");
  ASSERT_EQ(r.StringLiteral4(), "hello");
}

TEST(ComputedFieldsTest, DimensionIndex) {
  RecordWithComputedFields r;
  EXPECT_EQ(r.ArrayDimensionXIndex(), 0);
  EXPECT_EQ(r.ArrayDimensionYIndex(), 1);
  r.string_field = "y";
  EXPECT_EQ(r.ArrayDimensionIndexFromStringField(), 1);
  r.string_field = "missing";
  ASSERT_ANY_THROW(r.ArrayDimensionIndexFromStringField());
}

TEST(ComputedFieldsTest, DimensionCount) {
  RecordWithComputedFields r;
  EXPECT_EQ(r.ArrayDimensionCount(), 2);
  r.dynamic_array_field = {{1, 2, 3}, {4, 5, 6}};
  EXPECT_EQ(r.DynamicArrayDimensionCount(), 2);
  r.dynamic_array_field = {1, 2, 3};
  EXPECT_EQ(r.DynamicArrayDimensionCount(), 1);
}

TEST(ComputedFieldsTests, VectorSize) {
  RecordWithComputedFields r;
  EXPECT_EQ(r.VectorSize(), 0);
  r.vector_field = {1, 2, 3};
  EXPECT_EQ(r.VectorSize(), 3);

  EXPECT_EQ(r.FixedVectorSize(), 3);
}

TEST(ComputedFieldsTests, MapSize) {
  RecordWithComputedFields r;
  r.map_field = {{"hello", "bonjour"}, {"world", "monde"}};
  EXPECT_EQ(r.MapSize(), 2);
}

TEST(ComputedFieldsTests, ArraySize) {
  RecordWithComputedFields r;
  r.array_field = {{1, 2, 3}, {4, 5, 6}};
  ASSERT_EQ(r.ArraySize(), 6);
  ASSERT_EQ(r.ArrayXSize(), 2);
  ASSERT_EQ(r.ArrayYSize(), 3);
  ASSERT_EQ(r.Array0Size(), 2);
  ASSERT_EQ(r.Array1Size(), 3);

  ASSERT_EQ(r.ArraySizeFromIntField(), 2);
  r.int_field = 1;
  ASSERT_EQ(r.ArraySizeFromIntField(), 3);

  r.string_field = "x";
  ASSERT_EQ(r.ArraySizeFromStringField(), 2);
  r.string_field = "y";
  ASSERT_EQ(r.ArraySizeFromStringField(), 3);
  r.string_field = "missing";
  ASSERT_ANY_THROW(r.ArraySizeFromStringField());

  r.tuple_field.v1 = 1;
  ASSERT_EQ(r.ArraySizeFromNestedIntField(), 3);

  ASSERT_EQ(r.FixedArraySize(), yardl::size(r.fixed_array_field));
  ASSERT_EQ(r.FixedArrayXSize(), yardl::shape(r.fixed_array_field, 0));
  ASSERT_EQ(r.FixedArray0Size(), yardl::shape(r.fixed_array_field, 0));

  r.array_field_map_dimensions = {{1, 2, 3}, {4, 5, 6}};
  ASSERT_EQ(r.ArrayFieldMapDimensionsXSize(), 2);
}

TEST(ComputedFieldsTest, SwitchExpression) {
  RecordWithComputedFields r;
  r.optional_named_array = {{1, 2, 3}, {4, 5, 6}};
  ASSERT_EQ(r.OptionalNamedArrayLength(), 6);
  ASSERT_EQ(r.OptionalNamedArrayLengthWithDiscard(), 6);
  r.optional_named_array = {};
  ASSERT_EQ(r.OptionalNamedArrayLength(), 0);
  ASSERT_EQ(r.OptionalNamedArrayLengthWithDiscard(), 0);
  static_assert(std::is_same_v<decltype(r.OptionalNamedArrayLength()), yardl::Size>);
  static_assert(std::is_same_v<decltype(r.OptionalNamedArrayLengthWithDiscard()), yardl::Size>);

  r.int_float_union = 42;
  ASSERT_EQ(r.IntFloatUnionAsFloat(), 42.0f);
  static_assert(std::is_same_v<decltype(r.IntFloatUnionAsFloat()), float>);

  r.nullable_int_float_union = {};
  ASSERT_EQ(r.NullableIntFloatUnionString(), "null");
  r.nullable_int_float_union = 42;
  ASSERT_EQ(r.NullableIntFloatUnionString(), "int");
  r.nullable_int_float_union = 42.0f;
  ASSERT_EQ(r.NullableIntFloatUnionString(), "float");

  r.union_with_nested_generic_union = 42;
  ASSERT_EQ(r.NestedSwitch(), -1);
  ASSERT_EQ(r.UseNestedComputedField(), -1);
  r.union_with_nested_generic_union = basic_types::GenericRecordWithComputedFields<std::string, float>{"hi"};
  ASSERT_EQ(r.NestedSwitch(), 10);
  ASSERT_EQ(r.UseNestedComputedField(), 0);
  r.union_with_nested_generic_union = basic_types::GenericRecordWithComputedFields<std::string, float>{42.0f};
  ASSERT_EQ(r.NestedSwitch(), 20);
  ASSERT_EQ(r.UseNestedComputedField(), 1);

  basic_types::GenericRecordWithComputedFields<int, double> gr;
  gr.f1 = 42;
  ASSERT_EQ(gr.TypeIndex(), 0);
  gr.f1 = 42.0;
  ASSERT_EQ(gr.TypeIndex(), 1);
}

TEST(ComputedFieldsTest, Arithmetic) {
  RecordWithComputedFields r;
  ASSERT_EQ(r.Arithmetic1(), 3);
  static_assert(std::is_same_v<decltype(r.Arithmetic1()), int32_t>);
  ASSERT_EQ(r.Arithmetic2(), 11);
  ASSERT_EQ(r.Arithmetic3(), 13);

  r.array_field = {{1, 2, 3}, {4, 5, 6}};
  r.int_field = 1;
  ASSERT_EQ(r.Arithmetic4(), 5);
  ASSERT_EQ(r.Arithmetic5(), 3);

  ASSERT_EQ(r.Arithmetic6(), 3);

  ASSERT_EQ(r.Arithmetic7(), 49.0);
  static_assert(std::is_same_v<decltype(r.Arithmetic7()), double>);

  r.complexfloat32_field = {2.0f, 3.0f};
  ASSERT_EQ(r.Arithmetic8(), std::complex<float>(6.0f, 9.0f));

  ASSERT_EQ(r.Arithmetic9(), 2.2);
  static_assert(std::is_same_v<decltype(r.Arithmetic9()), double>);

  ASSERT_EQ(r.Arithmetic10(), 1e10 + 9e9);

  ASSERT_EQ(r.Arithmetic11(), -(5.3));
}

TEST(ComputedFieldsTest, Casting) {
  RecordWithComputedFields r;
  r.int_field = 42;
  ASSERT_EQ(r.CastIntToFloat(), 42.0f);
  static_assert(std::is_same_v<decltype(r.CastIntToFloat()), float>);

  r.float32_field = 42.9f;
  ASSERT_EQ(r.CastFloatToInt(), 42);
  static_assert(std::is_same_v<decltype(r.CastFloatToInt()), int32_t>);

  ASSERT_EQ(r.CastPower(), 49);
  static_assert(std::is_same_v<decltype(r.CastPower()), int32_t>);

  r.complexfloat32_field = {2.0f, 3.0f};
  r.complexfloat64_field = {2.0, 3.0};
  ASSERT_EQ(r.CastComplex32ToComplex64(), std::complex<double>(2.0, 3.0));
  ASSERT_EQ(r.CastComplex64ToComplex32(), std::complex<float>(2.0f, 3.0f));

  ASSERT_EQ(r.CastFloatToComplex(), std::complex<float>(66.6f, 0.0f));
  static_assert(std::is_same_v<decltype(r.CastFloatToComplex()), std::complex<float>>);
}

}  // namespace

File: matlab/test/ComputedFieldsTest.m
% Copyright (c) Microsoft Corporation.
% Licensed under the MIT License.

classdef ComputedFieldsTest < matlab.unittest.TestCase
    methods (Test)

        function testFieldAccess(testCase)
            r = test_model.RecordWithComputedFields();

            r.int_field = int32(42);
            testCase.verifyEqual(r.access_int_field(), int32(42));
            testCase.verifyEqual(r.access_other_computed_field(), r.access_int_field());

            r.string_field = "hello";
            testCase.verifyEqual(r.access_string_field(), "hello");

            r.tuple_field = test_model.MyTuple(1, 2);
            testCase.verifyEqual(r.access_tuple_field(), r.tuple_field);
            testCase.verifyEqual(r.access_nested_tuple_field(), r.tuple_field.v2);

            r.array_field = int32([[1; 2; 3], [4; 5; 6]]);
            testCase.verifyEqual(r.access_array_field(), r.array_field);
            testCase.verifyEqual(r.access_array_field_element(), r.array_field(2, 1));
            testCase.verifyEqual(r.access_array_field_element_by_name(), r.array_field(2, 1));

            r.vector_field = [1, 2, 3, 4];
            testCase.verifyEqual(r.access_vector_field, r.vector_field);
            testCase.verifyEqual(r.access_vector_field_element(), r.vector_field(2));
            r.vector_of_vectors_field = int32([[1; 2; 3], [4; 5; 6]]);
            testCase.verifyEqual(r.access_vector_of_vectors_field(), r.vector_of_vectors_field(3, 2));
            r.fixed_vector_of_vectors_field = int32([[1; 2; 3], [4; 5; 6]]);
            testCase.verifyEqual(r.access_fixed_vector_of_vectors_field(), r.fixed_vector_of_vectors_field(3, 2));

            r.map_field = dictionary(["hello", "world"], ["world", "bye"]);
            testCase.verifyEqual(r.access_map(), r.map_field);
            testCase.verifyEqual(r.access_map_entry(), "world");
            testCase.verifyEqual(r.access_map_entry_with_computed_field(), "world");
            testCase.verifyEqual(r.access_map_entry_with_computed_field_nested(), "bye");

            testCase.verifyError(@() r.access_missing_map_entry(), "MATLAB:dictionary:ScalarKeyNotFound");
        end

        function testLiterals(testCase)
            r = test_model.RecordWithComputedFields();
            testCase.verifyEqual(r.int_literal(), 42);
            testCase.verifyEqual(int64(r.large_negative_int64_literal()), -int64(0x4000000000000000));
            testCase.verifyEqual(uint64(r.large_u_int64_literal()), 0x8000000000000000);
            testCase.verifyEqual(r.string_literal(), "hello");
        end

        function testDimensionIndex(testCase)
            r = test_model.RecordWithComputedFields();

            testCase.verifyEqual(r.array_dimension_x_index(), 0);
            testCase.verifyEqual(r.array_dimension_y_index(), 1);

            r.string_field = "y";
            testCase.verifyEqual(r.array_dimension_index_from_string_field(), 1);

            r.string_field = "missing";
            testCase.verifyError(@() r.array_dimension_index_from_string_field(), "yardl:KeyError");
        end

        function testDimensionCount(testCase)
            r = test_model.RecordWithComputedFields();

            testCase.verifyEqual(r.array_dimension_count(), 2);

            r.dynamic_array_field = int32([[1; 2; 3], [4; 5; 6]]);
            testCase.verifyEqual(r.dynamic_array_dimension_count(), 2);
            r.dynamic_array_field = int32([1, 2, 3]);
            testCase.verifyEqual(r.dynamic_array_dimension_count(), 1);
        end

        function testVectorSize(testCase)
            r = test_model.RecordWithComputedFields();
            testCase.verifyEqual(r.vector_size(), 0);
            r.vector_field = [1, 2, 3, 4];
            testCase.verifyEqual(r.vector_size(), 4);

            testCase.verifyEqual(r.fixed_vector_size(), 3);
            testCase.verifyEqual(r.fixed_vector_of_vectors_size(), 2);
        end

        function testMapSize(testCase)
            r = test_model.RecordWithComputedFields();
            testCase.verifyEqual(r.map_size(), 0);
            r.map_field = dictionary(["hello", "bonjour"], ["world", "monde"]);
            testCase.verifyEqual(r.map_size(), 2);
        end

        function testArraySize(testCase)
            r = test_model.RecordWithComputedFields();
            r.array_field = int32([[1; 2; 3], [4; 5; 6]]);

            testCase.verifyEqual(r.array_size(), 6);
            testCase.verifyEqual(r.array_x_size(), 2);
            testCase.verifyEqual(r.array_y_size(), 3);
            testCase.verifyEqual(r.array_0_size(), 2);
            testCase.verifyEqual(r.array_1_size(), 3);

            testCase.verifyEqual(r.array_size_from_int_field(), 2);
            r.int_field = 1;
            testCase.verifyEqual(r.array_size_from_int_field(), 3);

            r.string_field = "x";
            testCase.verifyEqual(r.array_size_from_string_field(), 2);
            r.string_field = "y";
            testCase.verifyEqual(r.array_size_from_string_field(), 3);

            r.string_field = "missing";
            testCase.verifyError(@() r.array_size_from_string_field(), "yardl:KeyError");

            r.tuple_field.v1 = 1;
            testCase.verifyEqual(r.array_size_from_nested_int_field(), 3);
            testCase.verifyEqual(r.fixed_array_size(), 12);
            testCase.verifyEqual(r.fixed_array_x_size(), 3);
            testCase.verifyEqual(r.fixed_array_0_size(), 3);

            r.array_field_map_dimensions = int32([[1; 2; 3], [4; 5; 6]]);
            testCase.verifyEqual(r.array_field_map_dimensions_x_size(), 2);
        end

        function testSwitch(testCase)
            r = test_model.RecordWithComputedFields();
            r.optional_named_array = int32([[1; 2; 3], [4; 5; 6]]);
            testCase.verifyEqual(r.optional_named_array_length(), 6);
            testCase.verifyEqual(r.optional_named_array_length_with_discard(), 6);

            r.optional_named_array = yardl.None;
            testCase.verifyEqual(r.optional_named_array_length(), 0);
            testCase.verifyEqual(r.optional_named_array_length_with_discard(), 0);

            r.int_float_union = test_model.Int32OrFloat32.Int32(int32(42));
            testCase.verifyEqual(r.int_float_union_as_float(), single(42));
            r.int_float_union = test_model.Int32OrFloat32.Float32(single(42.9));
            testCase.verifyEqual(r.int_float_union_as_float(), single(42.9));

            r.nullable_int_float_union = yardl.None;
            testCase.verifyEqual(r.nullable_int_float_union_string(), "null");
            r.nullable_int_float_union = test_model.Int32OrFloat32.Int32(42);
            testCase.verifyEqual(r.nullable_int_float_union_string(), "int");
            r.nullable_int_float_union = test_model.Int32OrFloat32.Float32(42.9);
            testCase.verifyEqual(r.nullable_int_float_union_string(), "float");

            r.union_with_nested_generic_union = test_model.IntOrGenericRecordWithComputedFields.Int(42);
            testCase.verifyEqual(r.nested_switch(), -1);
            testCase.verifyEqual(r.use_nested_computed_field(), -1);

            g0 = basic_types.GenericRecordWithComputedFields(basic_types.T0OrT1.T0("hi"));
            r.union_with_nested_generic_union = ...
                test_model.IntOrGenericRecordWithComputedFields.GenericRecordWithComputedFields(g0);
            testCase.verifyEqual(r.nested_switch(), int32(10));
            testCase.verifyEqual(r.use_nested_computed_field(), int32(0));

            g1 = basic_types.GenericRecordWithComputedFields(basic_types.T0OrT1.T1(single(42.9)));
            r.union_with_nested_generic_union = ...
                test_model.IntOrGenericRecordWithComputedFields.GenericRecordWithComputedFields(g1);
            testCase.verifyEqual(r.nested_switch(), int32(20));
            testCase.verifyEqual(r.use_nested_computed_field(), int32(1));

            r.int_field = int32(10);
            testCase.verifyEqual(r.switch_over_single_value(), int32(10));


            gr = basic_types.GenericRecordWithComputedFields(basic_types.T0OrT1.T0(int32(42)));
            testCase.verifyEqual(gr.type_index(), 0);
            gr.f1 = basic_types.T0OrT1.T1(single(42.9));
            testCase.verifyEqual(gr.type_index(), 1);
        end

        function testArithmetic(testCase)
            r = test_model.RecordWithComputedFields();
            testCase.verifyEqual(r.arithmetic_1(), 3);
            testCase.verifyEqual(r.arithmetic_2(), 11);
            testCase.verifyEqual(r.arithmetic_3(), 13);

            r.array_field = int32([[1; 2; 3], [4; 5; 6]]);
            r.int_field = 1;
            testCase.verifyEqual(r.arithmetic_4(), 5);
            testCase.verifyEqual(r.arithmetic_5(), 3);

            testCase.verifyEqual(r.arithmetic_6(), 3.5);

            testCase.verifyEqual(r.arithmetic_7(), 49.0);

            r.complexfloat32_field = complex(2, 3);
            testCase.verifyEqual(r.arithmetic8(), single(complex(6, 9)));

            testCase.verifyEqual(r.arithmetic_9(), 2.2);
            testCase.verifyEqual(r.arithmetic_10(), 1e10 + 9e9);
        end

        function testCasting(testCase)
            r = test_model.RecordWithComputedFields();
            r.int_field = int32(42);
            testCase.verifyEqual(r.cast_int_to_float(), single(42));

            r.float32_field = 42.9;
            % NOTE: Matlab rounds floating point numbers when casting to integer
            testCase.verifyEqual(r.cast_float_to_int(), int32(43));

            testCase.verifyEqual(r.cast_power(), int32(49));

            r.complexfloat32_field = single(complex(2, 3));
            r.complexfloat64_field = complex(2, 3);
            testCase.verifyEqual(r.cast_complex32_to_complex64(), complex(2, 3));
            testCase.verifyEqual(r.cast_complex64_to_complex32(), single(complex(2, 3)));

            testCase.verifyEqual(r.cast_float_to_complex(), complex(single(66.6), 0));
        end

    end
end
